openhcl/underhill_core/src/dma_manager.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use memory_range::MemoryRange;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use user_driver::{memory::MemoryBlock, vfio::VfioDmaBuffer};
+
+pub struct GlobalDmaManager {
+    inner: Arc<Mutex<GlobalDmaManagerInner>>,
+}
+
+pub struct GlobalDmaManagerInner {
+    _physical_ranges: Vec<MemoryRange>,
+    _bounce_buffers_manager: Vec<MemoryRange>,
+    //clients: Mutex<Vec<Weak<DmaClient>>>,
+    //client_thresholds: Mutex<Vec<(Weak<DmaClient>, usize)>>,
+    dma_buffer_spawner: Box<dyn Fn(String) -> anyhow::Result<Arc<dyn VfioDmaBuffer>> + Send>,
+    clients: HashMap<String, Arc<DmaClient>>,
+}
+
+impl GlobalDmaManager {
+    pub fn new(
+        dma_buffer_spawner: Box<dyn Fn(String) -> anyhow::Result<Arc<dyn VfioDmaBuffer>> + Send>,
+    ) -> Self {
+        let inner = GlobalDmaManagerInner {
+            _physical_ranges: Vec::new(),
+            _bounce_buffers_manager: Vec::new(),
+            dma_buffer_spawner,
+            clients: HashMap::new(),
+        };
+
+        GlobalDmaManager {
+            inner: Arc::new(Mutex::new(inner)),
+        }
+    }
+
+    fn create_client_internal(
+        inner: &Arc<Mutex<GlobalDmaManagerInner>>,
+        pci_id: String,
+        device_name: String,
+    ) -> anyhow::Result<Arc<DmaClient>> {
+        let allocator = {
+            let manager_inner = inner
+                .lock()
+                .map_err(|_| anyhow::anyhow!("Failed to lock GlobalDmaManagerInner"))?;
+
+            // Access the page_pool and call its allocator method directly
+            (manager_inner.dma_buffer_spawner)(device_name)
+                .map_err(|e| anyhow::anyhow!("Failed to get DMA buffer allocator: {:?}", e))?
+        };
+
+        let client = DmaClient {
+            dma_manager_inner: inner.clone(),
+            dma_buffer_allocator: Some(allocator.clone()), // Set the allocator now
+        };
+
+        // Create an Arc<DmaClient>
+        let arc_client = Arc::new(client);
+
+        // Insert the client into the clients HashMap
+        let mut inner = inner.lock().expect("Failed to lock GlobalDmaManagerInner");
+        inner.clients.insert(pci_id, arc_client.clone());
+
+        Ok(arc_client) // Return the `Arc<Mutex<DmaClient>>`
+    }
+
+    pub fn get_client(&self, pci_id: &str) -> Option<Arc<DmaClient>> {
```

Comment:
again, why do we need this fn? shouldn't we just make a new client when requested?